compiler/rustc_typeck/src/check/generator_interior/drop_ranges.rs

```diff
@@ -0,0 +1,578 @@
+            trace!("Adding deferred edge from {} to {}", from, to);
+            self.add_control_edge(from, to)
+        });
+    }
+
+    pub fn drop_at(&mut self, value: HirId, location: usize) {
+        let value = self.hidx(value);
+        self.node_mut(location.into()).drops.push(value);
+    }
+
+    pub fn reinit_at(&mut self, value: HirId, location: usize) {
+        let value = match self.hir_id_map.get(&value) {
+            Some(value) => *value,
+            // If there's no value, this is never consumed and therefore is never dropped. We can
+            // ignore this.
+            None => return,
+        };
+        self.node_mut(location.into()).reinits.push(value);
+    }
+
+    pub fn propagate_to_fixpoint(&mut self) {
+        trace!("before fixpoint: {:#?}", self);
+        self.process_deferred_edges();
+        let preds = self.compute_predecessors();
+
+        trace!("predecessors: {:#?}", preds.iter_enumerated().collect::<BTreeMap<_, _>>());
+
+        let mut propagate = || {
+            let mut changed = false;
+            for id in self.nodes.indices() {
+                let old_state = self.nodes[id].drop_state.clone();
+                if preds[id].len() != 0 {
+                    self.nodes[id].drop_state = self.nodes[preds[id][0]].drop_state.clone();
+                    for pred in &preds[id][1..] {
+                        let state = self.nodes[*pred].drop_state.clone();
+                        self.nodes[id].drop_state.intersect(&state);
+                    }
+                } else {
+                    self.nodes[id].drop_state = if id.index() == 0 {
+                        BitSet::new_empty(self.num_values())
+                    } else {
+                        // If we are not the start node and we have no predecessors, treat
+                        // everything as dropped because there's no way to get here anyway.
+                        BitSet::new_filled(self.num_values())
+                    };
+                };
+                for drop in &self.nodes[id].drops.clone() {
+                    self.nodes[id].drop_state.insert(*drop);
+                }
+                for reinit in &self.nodes[id].reinits.clone() {
+                    self.nodes[id].drop_state.remove(*reinit);
+                }
+
+                changed |= old_state != self.nodes[id].drop_state;
+            }
+
+            changed
+        };
+
+        while propagate() {
+            trace!("drop_state changed, re-running propagation");
+        }
+
+        trace!("after fixpoint: {:#?}", self);
+    }
+
+    fn compute_predecessors(&self) -> IndexVec<PostOrderId, Vec<PostOrderId>> {
+        let mut preds = IndexVec::from_fn_n(|_| vec![], self.nodes.len());
+        for (id, node) in self.nodes.iter_enumerated() {
+            if node.successors.len() == 0 && id.index() != self.nodes.len() - 1 {
+                preds[<_>::from(id.index() + 1)].push(id);
```

Comment:
```suggestion
                preds[id.index() + 1].push(id);
```

Reply:
I believe the `Add` impl does what you want here. Regardless, `<_>::from` is an idiom i've not seen before, I'd prefer `PostOrderId::from`, makes the code easier to read, or else `(...).into()` perhaps.

Reply:
Huh, I tried a bunch of ways to make this work and apparently I missed the glaringly obvious `id + 1`.